### PR TITLE
Add permissions checking for editing users and permissions

### PR DESCRIPTION
**Pull request by EMcQ-BES**
### Changes:

- Add CRUDHandler to RouteHandler class structure
- Add base class for EditHandlers
- Restructure subdirectories for new class structure

### CODE REVIEW

File: packages/meditrak-server/src/routes/AccessRequests/assertAccessRequestPermissions.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ import {
 
 export const assertAccessRequestPermissions = async (accessPolicy, models, accessRequest) => {
   const results = await filterAccessRequestsByPermissions(accessPolicy, [accessRequest], models);
-  if (results.length === 0) {
+  if (results === 0) {
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/AccessRequests/assertAccessRequestPermissions.js]: Can you explain this one?

File: packages/meditrak-server/src/routes/editRecord.js [hunk 2]
```diff
@@ -47,6 +48,7 @@ const CUSTOM_RECORD_UPDATERS = {
   [TYPES.SURVEY]: editSurvey,
   [TYPES.SURVEY_SCREEN_COMPONENT]: editSurveyScreenComponent,
   [TYPES.ACCESS_REQUEST]: editAccessRequest,
+  [TYPES.USER_ENTITY_PERMISSION]: editUserEntityPermission,
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/editRecord.js]: Remove plz

File: packages/meditrak-server/src/app/addRoutesToApp.js [hunk 3]
```diff
@@ -159,6 +162,9 @@ export function addRoutesToApp(app) {
   /**
    * PUT routes
    */
+  app.put('(/v[0-9]+)/users/:recordId?', editUserAccounts);
+  app.put('(/v[0-9]+)/userEntityPermissions/:recordId?', editUserEntityPermissions);
+  app.put('(/v[0-9]+)/accessRequests/:recordId?', editAccessRequests);
```
Comment (edmofro) [inline packages/meditrak-server/src/app/addRoutesToApp.js]:
```suggestion
  app.put('(/v[0-9]+)/users/:recordId', editUserAccounts);
  app.put('(/v[0-9]+)/userEntityPermissions/:recordId', editUserEntityPermissions);
  app.put('(/v[0-9]+)/accessRequests/:recordId', editAccessRequests);
```

File: packages/meditrak-server/src/routes/AccessRequests/EditAccessRequests.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      ),
+    );
+  }
+
+  async editRecord() {
+    // Check Permissions
+    const accessRequest = await this.models.accessRequest.findById(this.recordId);
+    const accessRequestChecker = accessPolicy =>
+      assertAccessRequestPermissions(accessPolicy, this.models, accessRequest);
+    this.assertPermissions(assertAnyPermissions([assertBESAdminAccess, accessRequestChecker]));
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/AccessRequests/EditAccessRequests.js]:
```suggestion
    await this.assertPermissions(assertAnyPermissions([assertBESAdminAccess, accessRequestChecker]));
```

File: packages/meditrak-server/src/routes/EditHandler/EditHandler.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    await this.validate();
+    await this.editRecord();
+    respond(this.res, { message: `Successfully updated ${this.resource}` });
+  }
+
+  async validate() {
+    // dashboardReport, mapOverlay and dashboardGroup use different id formats and are the only endpoints
+    // which need to overwrite the validation functionality
+    // TODO remove when this task is done https://github.com/beyondessential/tupaia-backlog/issues/723
+    this.validateRecordExists();
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/EditHandler/EditHandler.js:29]:
```suggestion
    await this.validateRecordExists();
```
(which will wait for the promise to resolve, and then return undefined)
or
```suggestion
    return this.validateRecordExists();
```
(which will return the promise to be awaited by the caller, which will eventually resolve to the return value of validateRecordExists)

or even 
```suggestion
    return await this.validateRecordExists();
```
(which will wait for the promise to resolve, and then return the result of validateRecordExists)

File: packages/meditrak-server/src/routes/EditHandler/EditHandler.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  async editRecord() {
+    throw new Error('Any editHandler must implement editRecord()');
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/EditHandler/EditHandler.js:33]:
```suggestion
    throw new Error('Any EditHandler must implement editRecord()');
```

File: packages/meditrak-server/src/routes/EditHandler/EditHandler.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      .updateById(this.recordId, this.updatedFields);
+  }
+
+  async validateRecordExists() {
+    const validationCriteria = {
+      id: [constructRecordExistsWithId(this.database, this.recordType)],
+    };
+
+    const validator = new ObjectValidator(validationCriteria);
+    await validator.validate({ id: this.recordId }); // Will throw an error if not valid
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/EditHandler/EditHandler.js:48]: The style of await/return here should match the one you use above

File: packages/meditrak-server/src/routes/UserEntityPermissions/EditUserEntityPermissions.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    // Check Permissions
+    const userEntityPermission = await this.models.userEntityPermission.findById(this.recordId);
+    const userEntityPermissionChecker = accessPolicy =>
+      assertUserEntityPermissionPermissions(accessPolicy, this.models, userEntityPermission);
+    this.assertPermissions(
+      assertAnyPermissions([assertBESAdminAccess, userEntityPermissionChecker]),
+    );
+
+    // Update Record
+    this.updateRecord();
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/UserEntityPermissions/EditUserEntityPermissions.js]:
```suggestion
    return this.updateRecord();
```
Again you've got the three options of `await`, `return`, or `await return`. My preference is `return` - it doesn't have the redundant `await`, but does pass whatever the function being called returns straight back up (which can be handy)

**Review by edmofro — CHANGES_REQUESTED** [review]
Looking good! A few little suggestions

File: packages/meditrak-server/src/routes/AccessRequests/assertAccessRequestPermissions.js (hunk 1 above)
Comment (EMcQ-BES) [reply packages/meditrak-server/src/routes/AccessRequests/assertAccessRequestPermissions.js]: Forgot to update my branch after applying that fix in the previous PR, fixed.

File: packages/meditrak-server/src/routes/EditHandler/EditHandler.js (hunk 5 above)
Comment (EMcQ-BES) [reply packages/meditrak-server/src/routes/EditHandler/EditHandler.js]: I've updated all of these to `return`

**Review by edmofro — APPROVED** [review]

**Pull request merged**